### DISCUSSION TITLE
What is mitsuba.BSDF.eval()?

### DISCUSSION THREAD

**Discussion by Bambootree0818** [mitsuba-renderer/mitsuba3#1036] [Q&A] [answered]
Hello!
I am not sure about mitsuba.BSDF.eval().

The reference says "Evaluate the BSDF f(wi, wo) or its adjoint version f^{*}(wi, wo) and multiply by the cosine foreshortening term .".

This led me to believe that I could get the reflectance for the incident light. (e.g. reflectance for each wavelength of light).
But apparently not. The return value is a list of rgb.
What is this return value?

I would like to know what "Evaluate the BSDF" means.

**Comment by njroussel** [comment, marked as answer]
Hi @Bambootree0818 

This depends on which [variant](https://mitsuba.readthedocs.io/en/latest/src/key_topics/variants.html#part-3-color-representation) you are using. If you're using a `*_rgb` variant you will indeed be getting RGB values as a retrun value. In `*_spectral` variants you'll get the BSDF evaluated for the specific incident & outgoing directions at the given wavelengths (the wavelengths are specified in `SurfaceInteraction3f.wavelengths`).

**Reply by Bambootree0818** [reply]
Thank you very much!
The information you gave me is very close to what I need to know.

I tried and compiled mitsuba3 from source to use cuda_ad_spectral.
However, I encountered another problem.

I modified the BSDF tutorial in mitsuba3 a bit and tried to evaluate the BSDF at certain wavelengths in various directions.
I get no error but the program terminates with **si = dr.zeros(mi.SurfaceInteraction3f(ps,630))**. What am I doing wrong?
Here is the code
Wavelength is set at 630nm

``` Python
import mitsuba as mi
import drjit as dr

variants = mi.variants()
print(variants)

mi.set_variant('cuda_ad_spectral')
print(mi.variant())

bsdf = mi.load_dict({
    'type': 'principled',
    'base_color': {
            'type': 'rgb',
            'value': [1.0, 0.0, 0.0]
    },
    'metallic': 0.1,
    'specular': 1.0,
    'roughness': 0.5,
    'spec_tint': 0.0,
    'anisotropic': 0.0,
    'sheen': 0.0,
    'sheen_tint': 0.0,
    'clearcoat': 1.0,
    'clearcoat_gloss': 1.0,
    'spec_trans': 0.0
})

def sph_to_dir(theta, phi):
    st, ct = dr.sincos(theta)
    sp, cp = dr.sincos(phi)
    return mi.Vector3f(cp * st, sp * st, ct)

ps = mi.PositionSample3f()
si = dr.zeros(mi.SurfaceInteraction3f(ps,630))

si.wi = sph_to_dir(dr.deg2rad(0.0), 0)

res = 300
theta_o, phi_o = dr.meshgrid(
    dr.linspace(mi.Float, 0,     dr.pi,     res),
    dr.linspace(mi.Float, 0, 2 * dr.pi, 2 * res)
)
wo = sph_to_dir(theta_o, phi_o)

values = bsdf.eval(mi.BSDFContext(), si, wo)
print(values)
```

**Reply by Bambootree0818** [reply]
Hi! @njroussel 

I have corrected this code error and can now run it.
However, a question has arisen about the output results, so I will ask a new question in the issue.
Thank you for your help!
